locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,435 @@
+---
+title: Don't Block the Event Loop (or the Worker Pool)
+layout: docs.hbs
+---
+
+# Don't Block the Event Loop (or the Worker Pool)
+
+## Should you read this guide?
+If you're writing a server using Node, keep reading.
+If you're writing command-line scripts with Node, this guide might still come in handy, but it isn't crucial.
+
+## TL; DR
+Node.js runs JavaScript code in the Event Loop and handles I/O in a Worker Pool (the threadpool).
+Node scales well, sometimes better than more heavyweight approaches like Apache.
+The secret to Node's scalability is that it uses a small number of threads to handle many clients.
+If Node can make do with fewer threads, then it can spend more of your system's time and memory working on clients rather than on paying overheads for threads.
+
+Here's a good rule of thumb for keeping your Node server speedy:
+*Node is only fast when the work associated with each client at any given time is "small"*.
+
+This applies to work both on the Event Loop and on the Worker Pool.
+
+## Why should I avoid blocking the Event Loop and the Worker Pool?
+Node uses a small number of threads to handle many clients.
+While a thread is working on behalf of one client, it can't work for any other clients.
+
+This provides two motivations for blocking neither the Event Loop nor the Worker Pool.
+Since both the Event Loop and the Workers respond to events, we will refer to these threads generally as *Event Handlers* or *Event-Handler Threads*.
+All threads in Node are Event Handlers: one Event Loop and several Workers.
+
+1. Performance: If you regularly perform heavyweight tasks on an Event Handler, the *throughput* (requests/second) of your server will suffer.
+2. Security: If it is possible that for certain input you might block an Event Handler, a malicious client might be able to figure out the "evil input", make your code block, and keep you from working on other clients. This would be a Denial of Service attack.
+
+## A quick review of Node's architecture
+
+Node uses the Event-Driven Architecture: it has an Event Loop for orchestration and a Worker Pool for expensive tasks.
```

Comment:
file i/o isn't expensive, no more so than network i/o, it just doesn't have good completion APIs at the O/S level to allow it to be done on the main loop, so threads are needed so that i/o can be done with blocking system calls

Reply:
My understanding of the libuv loop (which is limited to Linux) is that the trouble is not the completion APIs but rather the ability of the OS to say that the file descriptor is ready and thus not to block when you do I/O on it. libuv only sends network traffic when it will not block in doing so, but the OS will not guarantee that read/write on a file descriptor will be non-blocking.

Based on this understanding, my claim is correct - the I/O done on the event loop is always cheap *from the event loop's perspective* because the OS guarantees it to be non-blocking.

Am I missing something here?

Reply:
Maybe. The actual problem that is that you can't poll/epoll/select/kqueue a file descriptor for readiness, and so we can't do non-blocking reads or writes. And I guess you could call a blocking call "expensive", but network i/o can be done in thread pools with blocking calls, and lots of languages do exactly that. node doesn't, because network i/o can be non-blocking on unix, whereas file i/o can't (without the async i/o posix extensions, but they are poorly supported, and while windows has better async i/o for files, early tests by libuv authors showed that it would randomly block, it was to flaky to be useful).

Reply:
Last I checked, posix async I/O is just a threadpool in libc, no different from the libuv threadpool. I've prototyped Node with kernel AIO to see if things had improved since [2011](https://github.com/nodejs/node-v0.x-archive/issues/1446), but the O_DIRECT restriction remains and its costs are catastrophic.

Anyway, I'll flesh out this discussion a bit more.

Reply:
aio was intended to be implemented in kernel and thus to rock... but hasn't happened yet, the saga is long, some info here: https://lwn.net/Articles/724198/. There is various user-space emulation using blocking I/O in a thread pool....  so, yeah, uv doesn't use that, there is no point.